lib.php

```diff
@@ -25,133 +25,129 @@
+                    tool_abconfig_execute_command_array($experiments[$experiment]['conditions'][$condition]['commands'], $experiments[$experiment]['shortname']);
+                }
             }
         }
-    }
 
-    $commandarray = array();
+        $commandarray = array();
 
-    // First, Build a list of all commands that need to be executed
-    // Every experiment that is per request
-    $compare = $DB->sql_compare_text('request', strlen('request'));
-    $records = $DB->get_records_sql("SELECT * FROM {tool_abconfig_experiment} WHERE scope = ? AND enabled=1", array($compare));
+        // First, Build a list of all commands that need to be executed
 
-    foreach ($records as $record) {
+        // Start with request scope
+        $requestexperiments = $manager->get_active_request();
 
-        // Make admin immune unless enabled for admin
-        if (is_siteadmin()) {
-            if ($record->adminenabled == 0) {
-                continue;
+        foreach ($requestexperiments as $record) {
+
+            // Make admin immune unless enabled for admin
+            if (is_siteadmin()) {
+                if ($record['adminenabled'] == 0) {
+                    continue;
+                }
             }
-        }
 
-        $conditionrecords = $DB->get_records('tool_abconfig_condition', array('experiment' => $record->id));
+            $conditionrecords = $record['conditions'];
 
-        // Remove all conditions that contain the user ip in the whitelist
-        $crecords = array();
+            // Remove all conditions that contain the user ip in the whitelist
+            $crecords = array();
 
-        foreach ($conditionrecords as $conditionrecord) {
-            $iplist = $conditionrecord->ipwhitelist;
-            if (!remoteip_in_list($iplist)) {
-                array_push($crecords, $conditionrecord);
+            foreach ($conditionrecords as $conditionrecord) {
+                $iplist = $conditionrecord['ipwhitelist'];
+                if (!remoteip_in_list($iplist)) {
+                    array_push($crecords, $conditionrecord);
+                }
             }
-        }
 
-        // Increment through conditions until one is selected
-        $condition = '';
-        $num = rand(1, 100);
-        $prevtotal = 0;
-        foreach ($crecords as $crecord) {
-            // If random number is within this range, set condition and break, else increment total
-            if ($num > $prevtotal && $num <= ($prevtotal + $crecord->value)) {
-                $commandarray[$record->shortname] = $crecord->commands;
-
-                // Do not select any more conditions
-                break;
-            } else {
-                // Not this record, increment lower bound, and move on
-                $prevtotal += $crecord->value;
+            // Increment through conditions until one is selected
+            $condition = '';
+            $num = rand(1, 100);
```

Comment:
unrelated, but should this be rand(0, 100); not 1?

Reply:
I just did some checks, if it is rand(0,100), on the next line, `$num > $prevtotal` this check can fail if it rolls a 0, which would exclude the first condition set if it was meant to fire. With this current logic, rand(1,100) should be correct